index.bs

```diff
@@ -257,28 +264,112 @@ spec:css-syntax-3;
   This document depends on the Infra Standard for a number of foundational concepts used in its
   algorithms and prose [[!INFRA]].
 
+
+  ### Infrastructure Algorithms ### {#sctn-infra-algorithms}
+
+  <!-- NOTE: '<h4>' maps to the same heading level as three hashs '###', thus using '<h5>' for the next section -->
+  <h5 id="algorithm-same-origin-with-ancestors" algorithm>Same-Origin with its Ancestors</h5>
+
   An [=environment settings object=] (|settings|) is <dfn noexport>same-origin with its
   ancestors</dfn> if the following algorithm returns `true`:
 
-  1.  If |settings| has no [=environment settings object/responsible document=],
-      return `false`.
+  <ol class="algorithm">
+    1.  If |settings| has no [=environment settings object/responsible document=],
+        return `false`.
+
+    2.  Let |document| be |settings|' [=environment settings object/responsible document=].
 
-  2.  Let |document| be |settings|' [=environment settings object/responsible document=].
+    3.  If |document| has no [=document/browsing context=], return `false`.
 
-  3.  If |document| has no [=document/browsing context=], return `false`.
+    4.  Let |origin| be |settings|' [=environment settings object/origin=].
 
-  4.  Let |origin| be |settings|' [=environment settings object/origin=].
+    5.  Let |current| be |document|'s [=document/browsing context=].
 
-  5.  Let |current| be |document|'s [=document/browsing context=].
+    6.  While |current| has a [=parent browsing context=]:
+
+        1.  Set |current| to |current|'s [=parent browsing context=].
+
+        2.  If |current|'s [=active document=]'s [=origin=] is not [=same origin=] with |origin|,
+            return `false`.
+
+    7.  Return `true`.
+  </ol>
+
+  ### Registry: Credential Type to Options Member Identifier Mappings ### {#sctn-registry-cred-type-to-options-member-identifier}
+
+  This registry maps [=credential types=] (i.e., {{Credential/[[type]]}} values) to the
+  [=dictionary member=] [=identifier=] used in
+  {{CredentialCreationOptions}} and {{CredentialRequestOptions}} (i.e., "options dictionaries")
+  by their specifications.
+
+  Note: This registry is used by the [=relevant credential interface objects=] algorithm.
+
+
+  <table class="complex data longlastcol">
+    <tr>
+        <th>Credential Type<br>
+        (in alphabetical order)</th>
+        <th>Options Member Identifier</th>
```

Comment:
Add a column for the appropriate interface object (e.g. `{{PublicKeyCredential}}`) so that the algorithm doesn't have to iterate over all interface objects in the global.

Reply:
Hm. Perhaps you can explain further? (I must be missing something):

I do not understand how the registry entry having the "appropriate interface object" will obviate needing to look through the global object's interface objects to find all instances of that object type?

Or, are you saying we could then (in this spec) say something like the below, specifically step (c):
> a. let _optionsKeys_ be a new map populated with the result of [getting the keys](https://infra.spec.whatwg.org/#map-getting-the-keys) of _options_.
> b. let _appropriateInterfaceObject_ be the result of: look up each individual _optionkey_ in _optionsKeys_ in the registry, and if there is a matching entry, return the value in the "appropriate interface object" column.
> c. let _relevant interface objects_ be the set of interface objects gathered from _settings_' global object whose type match the value of _appropriateInterfaceObject_.

...?  (IMV, the above seems essentially just as complicated as the present alg and in essence continues to iterate thru global's interface objects)

Reply:
> find all instances of that object type

I believe there's only one instance of each interface object, e.g. only one instance of the `{{PublicKeyCredential}}` interface, so you wouldn't need to do step c. Instead, you'd presumably `append the value in the "appropriate interface object" column to the relevant interface objects set.` in step b.

Reply:
Yep. If we want to be extra precise, we could talk about

> the [interface object](https://webidl.spec.whatwg.org/#dfn-interface-object) on _settings_' global object for the [=Credential Type Registry/Credential Interface=] whose [=Credential Type Registry/Options Member Identifier=] is |optionkey|

Reply:
OK, then IIUC, this (adding the "appropriate interface object" column and altering this algorithm) will be something to do in the (not yet created) PR resolving issue #184.

Reply:
> Add a column for the appropriate interface object (e.g. {{PublicKeyCredential}}) so that the algorithm doesn't have to iterate over all interface objects in the global.

[ ... ]
> I believe there's only one instance of each interface object, e.g. only one instance of the {{PublicKeyCredential}} interface ...

> If we want to be extra precise, we could talk about
>> the [interface object](https://webidl.spec.whatwg.org/#dfn-interface-object) on _settings_' global object for the [=Credential Type Registry/Credential Interface=] whose [=Credential Type Registry/Options Member Identifier=] is |optionkey|

Hm.  Ok, IIUC, we can change the [relevant credential interface objects](https://www.w3.org/TR/credential-management-1/#credentialrequestoptions-relevant-credential-interface-objects) alg to iterate through the the keys of the given |options| and append any existing interface object on settings' global object for the [=Credential Type Registry/Credential Interface=] whose [=Credential Type Registry/Options Member Identifier=] is |optionkey| (_if_ such an object exists, _and_ has the proper `[[type]]` slot value ...?)

I.e., IIUC, it seems credman is crafted to allow for the given options dictionary to contain more than one key/value pair, i.e., one or more (unique?) values from the [=Credential Type Registry/Options Member Identifier=] column.

If the above understanding is correct, then we can rewrite the [relevant credential interface objects](https://www.w3.org/TR/credential-management-1/#credentialrequestoptions-relevant-credential-interface-objects) alg to be like (tho, see issues/questions below):

1. Let |settings| be the [=current settings object=].

1. Let |relevant interface objects| be an [=set/empty=] [=set=].

1. Let |optionsKeys| be a [=ordered set=] whose [=list/items=] are the result of [map/getting the keys=] of |options|.

1. [list/iterate|For each=] |optionKey| of |optionsKeys|:

    1. Let |credentialInterfaceIdentifier| be the [=credential type registry/Credential Interface Object Identifier=] (found using the [[#sctn-cred-type-registry|Credential Type Registry]]) whose [=credential type registry/Options Member Identifier=] is |optionKey|.

    1. Let |credentialTypeValue| be the [=credential type registry/Credential Type=]  whose [=credential type registry/Options Member Identifier=] is |optionKey|.

    1. If an [=interface object=] whose [=identifier=] is |credentialInterfaceIdentifier| exists on |settings|' [=environment settings object/global object=], let |credentialInterfaceObject| be that [=interface object=], otherwise [=iteration/continue=].

    1. If |credentialInterfaceObject|'s {{[[type]]}} slot's value does not match |credentialTypeValue|, [=iteration/continue=].

    1. Append |credentialInterfaceObject| to |relevant interface objects|.

1.  Return |relevant interface objects|.

Issues/Questions: 

A. This rewritten alg seems similarly complex (to me) as the present "[iterate over global's interfaces](https://www.w3.org/TR/credential-management-1/#credentialrequestoptions-relevant-credential-interface-objects)" approach, thus I'm uncertain whether this is better?

B. wrt @jyasskin's suggested approach of:
> the [interface object](https://webidl.spec.whatwg.org/#dfn-interface-object) on _settings_' global object for the [=Credential Type Registry/Credential Interface=] whose [=Credential Type Registry/Options Member Identifier=] is |optionkey|

...I added obtaining intermediate values as well as explicit correctness checks because it is not clear to me that the suggested approach would inherently properly guide an implementor.

C. Is the above re-written alg more-or-less correct?  If not, what am I misunderstanding?

Reply:
I think https://github.com/w3c/webappsec-credential-management/compare/jeffh-fix-147-options-member-name-and-cred-type...jyasskin:table-based-relevant-credentials works here. The rewritten alg seems correct, but more complicated than necessary.

Reply:
Thanks for the suggestion, I merged it in: 32c086c